lib/internals.js

```diff
@@ -0,0 +1,930 @@
+          })
+        ];
+        entrypoint.manifest = {
+          root: self.distFolderName,
+          files: {
+            js: [ manifest.file ],
+            css,
+            assets,
+            imports,
+            dynamicImports
+          },
+          src: {
+            js: [ manifest.src ]
+          },
+          devServer: self.hasDevServerUrl(entrypoint.type)
+        };
+        result.push(entrypoint);
+      }
+
+      function defaultConvertFn(ref) {
+        return ref;
+      }
+      function getFiles({
+        manifest, entry, data, sources, target, convertFn = defaultConvertFn
+      }, acc = [], seen = {}) {
+        if (Array.isArray(data)) {
+          acc.push(...data.map(convertFn));
+        }
+        for (const source of sources) {
+          if (!Array.isArray(entry?.[source])) {
+            continue;
+          }
+          entry[source].forEach(ref => {
+            if (seen[`${source}-${ref}`]) {
+              return;
+            }
+            seen[`${source}-${ref}`] = true;
+            manifest[ref] && getFiles({
+              manifest,
+              entry: manifest[ref],
+              data: manifest[ref][target],
+              sources,
+              target,
+              convertFn
+            }, acc, seen);
+          });
+        }
+        return acc;
+      }
+
+      return result;
+    },
+
+    // Accepts an entrypoint and optional files object and returns a manifest-like object.
+    // This handler is used in the initializing phase of the build process.
+    // In scenarios where the module build is not tirggered at all
+    // (e.g. boot in production),
+    // the core system will use its own saved manifest to identify the files that has to
+    // be injected in the browser. This manifest is mostly used in development (especially
+    // the `devServer` property) when a build for given entrypoint is not triggered
+    // (because this entrypoint is served by the dev server).
+    toManifest(entrypoint, files) {
+      if (entrypoint.type === 'bundled') {
+        const result = {
+          root: '',
+          files: {
+            js: files?.js || [],
+            css: files?.css || [],
+            assets: [],
+            imports: [],
+            dynamicImports: []
+          },
+          // Bundled entrypoints are not served by the dev server.
+          src: null,
+          devServer: false
+        };
+        if (result.files.js.length || result.files.css.length) {
+          return result;
+        }
+        return null;
+      }
+      return {
+        root: self.distFolderName,
+        files: {
+          js: [],
+          css: [],
+          assets: [],
+          imports: [],
+          dynamicImports: []
+        },
+        // This can be extended, for now we only support JS entries.
+        // It's used to inject the entrypoint into the HTML.
+        src: {
+          js: [ path.join(self.buildSourceFolderName, `${entrypoint.name}.js`) ]
```

Comment:
can we guarantee that the entrypoint name will not already contain .js at the end, or be something else?

Reply:
Yes, guaranteed.